sstables/sstables.cc

```diff
@@ -3398,15 +3400,23 @@ future<> sstable::s3_storage::change_state(const sstable& sst, sstring to, gener
 }
 
 future<> sstable::s3_storage::wipe(const sstable& sst) noexcept {
-    auto& sys_ks = sst.manager().system_keyspace();
-
-    co_await sys_ks.sstables_registry_update_entry_status(_location, sst.generation(), status_removing);
-
     co_await coroutine::parallel_for_each(sst._recognized_components, [this, &sst] (auto type) -> future<> {
         co_await _client->delete_object(make_s3_object_name(sst, type));
     });
+}
+
+future<> sstable::s3_storage::delete_with_system_keyspace(std::vector<shared_sstable> ssts) {
+    co_await coroutine::parallel_for_each(ssts, [] (shared_sstable sst) -> future<> {
+        const s3_storage* storage = dynamic_cast<const s3_storage*>(&sst->get_storage());
+        if (!storage) {
+            on_fatal_internal_error(sstlog, "Atomically deleted sstables must be of same storage type");
+        }
 
-    co_await sys_ks.sstables_registry_delete_entry(_location, sst.generation());
+        auto& sys_ks = sst->manager().system_keyspace();
+        co_await sys_ks.sstables_registry_update_entry_status(storage->_location, sst->generation(), status_removing);
+        co_await sst->unlink();
+        co_await sys_ks.sstables_registry_delete_entry(storage->_location, sst->generation());
+    });
```

Comment:
I think sys_ks.sstables_registry_delete_entry() call should be moved into wipe() for symmetry. Not all deletion can go through delete_atomically(), so we'll leave stale entries for deleted objects.

Reply:
But that's one of the points of the PR -- the .wipe() method of the storage doesn't need to be atomic for any storage type. It's already non-atomic for filesystem storage, but happened to be such for s3. If not all deletion can go through delete_atomically, then we're in trouble already, as filesystem_storage::wipe() is non-atomic (and never was)

Reply:
if compaction is running, creates 1 sstable which is still in unsealed state, there's an entry for it in sstable registry table right? then that compaction fails, meaning that it will only discard that sstable using sstable::unlink(). there are a few places (under similar scenario) where we don't need atomicity and call unlink directly. will we leave stale entries? deleting entry in sstable registry seem unrelated to atomic deletion and it seems to me that it could live in wipe() for symmetry. we add a entry in open(), we remove the entry in wipe().

Reply:
> if compaction is running, creates 1 sstable which is still in unsealed state, there's an entry for it in sstable registry table right? 

Should be. The entry is placed into registry on sstable::open(). Similarly, in open() the TemporaryTOC file is created for disk-backed sstables.

> then that compaction fails, meaning that it will only discard that sstable using sstable::unlink(). 

What if the sstable is on-disk and .wipe() crashes in the middle?

> there are a few places (under similar scenario) where we don't need atomicity and call unlink directly.

Then I misundertand what "atomically" means wrt sstables. My thought was that atomically means in a way that would unlink them even in case of crash in the middle.

> will we leave stale entries? 

I would say ... yes, but that's not the problem of this PR. The sstables you describe would be in "creating" state. Issue #13024 should take care of dangling stuff.

> deleting entry in sstable registry seem unrelated to atomic deletion and it seems to me that it could live in wipe() for symmetry. we add a entry in open(), we remove the entry in wipe().

Reply:
Ah there's two types of atomicity. Atomicity with regards to a single sstable, where we want a sstable to be completely removed or not. And atomicity with regards to many sstables, where we want sstables A and B to be removed together, to avoid data ressurection. I talked more about it in my last comment in the third patch.

Reply:
delete_atomically() should implement t both types of atomicity. 

whereas unlink() should implement the atomic deletion for a single sstable, meaning it makes sense to remove the sstable registry entry in the wipe() call.

if you look at sstable::filesystem_storage::wipe(), you will see it calls remove_by_toc_name(), which plays with temporary TOC component to guarantee a single sstable is completely removed wrt its components

Reply:
Ah, I see. Yes, then the entry update should happen in wipe as it is now, but then we're in ... trouble? Consider two sstables are asked to be deleted atomically. How would that work wrt ownership table?

Consider we first update status for both into "deleting", then call wipe(). In that case failure to update the status on the 2nd table would result in 1st being dropped after reboot and 2nd continuing being alive.

Reply:
Yes, even the split is not enough. We'd need batch updates. My concern about batches is that compactions like reshape have hundreds of entries deleted together. If you have 96 shards, you'll have 96 batches, with hundreds of operations each, happening roughly at the same time. The alternative is to implement deletion logs in S3, perhaps storing an entry for the log in sstables_registry and reusing existing fields like format and version to identify it.

Reply:
OK, took the note #13567